features/timeseries/src/main/java/org/opennms/netmgt/timeseries/meta/TimeSeriesMetaDataDao.java

```diff
@@ -69,34 +69,35 @@
 
     private final DataSource dataSource;
 
-    private final Map<String, Map<String, String>> cache; // resourceId, Map<name, value>
+    private final Cache<String, Map<String, String>> cache; // resourceId, Map<name, value>
 
     @Autowired
     public TimeSeriesMetaDataDao(final DataSource dataSource,
                                  @Named("timeseries.metadata.cache_size") long cacheSize,
                                  @Named("timeseries.metadata.cache_duration") long cacheDuration) {
         this.dataSource = dataSource;
-        Cache<String, Map<String, String>> cache = CacheBuilder.newBuilder()
+        this.cache = CacheBuilder.newBuilder()
                 .maximumSize(cacheSize)
                 .expireAfterWrite(cacheDuration, TimeUnit.SECONDS) // to make sure cache and db are in sync
                 .build();
-        this.cache = cache.asMap();
     }
 
-    public void store(final Collection<MetaData> metaDataCollection) throws SQLException {
+    public void store(final Collection<MetaData> metaDataCollection) throws SQLException, ExecutionException {
         Objects.requireNonNull(metaDataCollection);
-        Set<MetaData> uncached = new HashSet<>();
+        Set<MetaData> writeToDb = new HashSet<>();
 
         // find all MetaData that is not present in the cache
         for(MetaData meta : metaDataCollection) {
-            if(!Optional
-                    .ofNullable(cache.get(meta.getResourceId()))
-                    .map(entry -> entry.get(meta.getName()))
-                    .isPresent()) {
-                uncached.add(meta);
+            Map<String, String> attributesForResource = cache.get(meta.getResourceId(), HashMap::new);
+            if(attributesForResource.get(meta.getName()) == null) {
```

Comment:
```suggestion
            Map<String, String> attributesForResource = cache.getIfPresent(meta.getResourceId());
            if(attributesForResource != null && attributesForResource.contains(meta.getName())) {
```

To save the allocation

Reply:
:+1: